### ISSUE TITLE
DoS vulnerability via obsolete "css-what" version

### ISSUE THREAD

**Issue by khitrenovich** [cheeriojs/cheerio#1924]
There is fresh a security advisory on `css-what` in `cheerio` v0.22 - https://www.npmjs.com/advisories/1754.

Audit report - 
```
yarn audit v1.22.10
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Denial of Service                                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ css-what                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=5.0.1                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ cheerio                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ cheerio > css-select > css-what                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1754                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

Note that `css-what` version that is currently in use is v2.1 and forced bump to v5 breaks `cheerio` - so that's not an option.

**Comment by fb55** [comment]
This issue only affects css-what@4.0.0 and higher. I am trying to get the CVE updated to reflect this.

For now, I have just published cheerio@1.0.0-rc.10, which includes the version update.

**Comment by khitrenovich** [comment]
@fb55 I hope you'll manage to update the CVE. We use fragments in our project, so v1.0 is not a drop-in replacement.

**Comment by fb55** [comment]
Sent an email, let's see how long this will take.

Out of curiosity: What do you mean with fragments? The module itself hasn't changed too much, so curious to hear how we can improve the adoption story.

**Comment by khitrenovich** [comment]
> Sent an email, let's see how long this will take.

Last time I had to do something like that, it was a matter of hours for the advisory team to take care.

> Out of curiosity: What do you mean with fragments? The module itself hasn't changed too much, so curious to hear how we can improve the adoption story.

We are manipulating HTML templates, which can be each full document (with `<html>`/`<head>`/`<body>`) or fragments (single `<div>` elements, for example). We "patch" certain elements with either generated content or content from another templates. In v0.x, `$.html()` returned either fragment or the full document, depending on what was initially loaded. In v1.0 we always get back full document wrapped with `<html>` etc.

That's what I see in the unit test failure, for example - 
```
       -   "template": "<div id=\"content\"><div>email content</div></div>",
       +   "template": "<html><head></head><body><div id=\"content\"><div>email content</div></div></body></html>",
```

**Comment by fb55** [comment]
> In v1.0 we always get back full document wrapped with <html> etc.

All you'll have to do to disable this behavior is set the `isDocument` boolean of `load` to false:

```js
const $ = cheerio.load(htmlString, null, false)
```

**Comment by davidowens** [comment]
> This issue only affects css-what@4.0.0

@fb55 can you point me to where you found that?

**Comment by fb55** [comment]
> @fb55 can you point me to where you found that?

I am the maintainer of `css-what`. The bug was introduced in https://github.com/fb55/css-what/commit/63cb253baf66a64094de8ecc167d41b4d4dee90b

**Comment by davidowens** [comment]
awesome thank you!

**Comment by khitrenovich** [comment]
> All you'll have to do to disable this behavior is set the `isDocument` boolean of `load` to false:
> 
> ```js
> const $ = cheerio.load(htmlString, null, false)
> ```

Great, this is very helpful! I think it is worth explicitly documenting that flag in "migrating from v0.x" section.

**Comment by curvedriver** [comment]
> This issue only affects css-what@4.0.0 and higher. I am trying to get the CVE updated to reflect this.
> 
> For now, I have published cheerio@1.0.0-rc.10, which includes the version update.

FYI, I requested an update of the npm advisory so that the version list is equal to the CVE and `npm audit` no longer reports the security vulnerability.

https://npmjs.com/advisories/1754

**Comment by fb55** [comment]
Thanks @curvedriver! Closing this as everything seems to be taken care of.
